### PR TITLE
Feat: When using Airflow source the default catalog from Airflow variables

### PR DESCRIPTION
**Pull request by izeigerman**

### CODE REVIEW

**Review by tobymao — APPROVED** [review]

File: sqlmesh/core/config/scheduler.py [hunk 1]
```diff
@@ -96,6 +95,9 @@ def create_plan_evaluator(self, context: Context) -> PlanEvaluator:
             notification_target_manager=context.notification_target_manager,
         )
 
+    def get_default_catalog(self, context: Context) -> t.Optional[str]:
+        return context.engine_adapter.default_catalog
```
Comment (izeigerman) [inline sqlmesh/core/config/scheduler.py:99]: It occurred to me now: shall we raise on the client side as well when the variable is not set?

Comment (eakmanrq) [reply sqlmesh/core/config/scheduler.py:99]: I updated this for the Airflow schedulers to raise. Builtin still supports not having a default catalog.

File: docs/integrations/engines/bigquery.md [hunk 2]
```diff
@@ -52,6 +52,7 @@ By default, the connection ID is set to `sqlmesh_google_cloud_bigquery_default`,
 ```python linenums="1"
 sqlmesh_airflow = SQLMeshAirflow(
     "bigquery",
+    default_catalog="<project name>",
```
Comment (izeigerman) [inline docs/integrations/engines/bigquery.md:55]: [Nit] I think in BQ they refer to it as project id?

File: sqlmesh/core/config/scheduler.py [hunk 3]
```diff
@@ -296,7 +306,14 @@ def create_plan_evaluator(self, context: Context) -> PlanEvaluator:
         )
 
     def get_default_catalog(self, context: Context) -> t.Optional[str]:
-        return None
+        default_catalog = self.get_client(context.console).get_variable(
```
Comment (izeigerman) [inline sqlmesh/core/config/scheduler.py:309]: I wonder whether we can DRY up this block a little

Comment (izeigerman) [reply sqlmesh/core/config/scheduler.py:309]: For context: `MWAAClient` and `AirflowClient` share a common base class with `get_variable` available in it. So we can easily write a function for this.

File: sqlmesh/schedulers/airflow/integration.py [hunk 4]
```diff
@@ -101,6 +103,15 @@ def __init__(
         self._plan_application_dag_ttl = plan_application_dag_ttl
         self._external_table_sensor_factory = external_table_sensor_factory
         self._generate_cadence_dags = generate_cadence_dags
+        self.set_default_catalog(default_catalog)
```
Comment (izeigerman) [inline sqlmesh/schedulers/airflow/integration.py:106]: Should we perhaps do this in `dags` and avoid side effects in the construtor?

File: sqlmesh/schedulers/airflow/integration.py [hunk 5]
```diff
@@ -101,6 +103,15 @@ def __init__(
...
+
+    @classmethod
+    def set_default_catalog(cls, default_catalog: str) -> None:
+        current_value = Variable.get(common.DEFAULT_CATALOG_VARIABLE_NAME, default_var=None)
+        if not current_value:
+            Variable.set(common.DEFAULT_CATALOG_VARIABLE_NAME, default_catalog)
```
Comment (izeigerman) [inline sqlmesh/schedulers/airflow/integration.py:112]: As far as I can tell `set` automatically deletes the key if it exists. Can we just set unconditionally?

Comment (izeigerman) [reply sqlmesh/schedulers/airflow/integration.py:112]:
```
    def set(
        cls,
        key: str,
        value: Any,
        description: Optional[str] = None,
        serialize_json: bool = False,
        session: Session = None,
    ):
        """
        Sets a value for an Airflow Variable with a given Key.
        This operation will overwrite an existing variable.

        :param key: Variable Key
        :param value: Value to set for the Variable
        :param description: Description of the Variable
        :param serialize_json: Serialize the value to a JSON string
        :param session: SQL Alchemy Sessions
        """
        # check if the secret exists in the custom secrets backend.
        cls.check_for_write_conflict(key)
        if serialize_json:
            stored_value = json.dumps(value, indent=2)
        else:
            stored_value = str(value)

        Variable.delete(key, session=session)
        session.add(Variable(key=key, val=stored_value, description=description))
        session.flush()
```

Comment (eakmanrq) [reply sqlmesh/schedulers/airflow/integration.py:112]: I got an error when I tried that

Comment (izeigerman) [reply sqlmesh/schedulers/airflow/integration.py:112]: Hm, based on the code above I don't see how this could happen

Comment (izeigerman) [reply sqlmesh/schedulers/airflow/integration.py:112]: Ah, probably this:
```
cls.check_for_write_conflict(key)
```

Comment (izeigerman) [reply sqlmesh/schedulers/airflow/integration.py:112]: hm, no, it should only matter if there's a key collision with the secretes backend. What error did you get?

Comment (eakmanrq) [reply sqlmesh/schedulers/airflow/integration.py:112]:
```
[2023-12-20 20:03:28,923: ERROR/ForkPoolWorker-1] Failed to import: /opt/airflow/dags/sqlmesh_integration.py
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1803, in _execute_context
    cursor, statement, parameters, context
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 719, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "variable_key_key"
DETAIL:  Key (key)=(sqlmesh_default_catalog) already exists.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/models/dagbag.py", line 318, in parse
    loader.exec_module(new_module)
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/opt/airflow/dags/sqlmesh_integration.py", line 6, in <module>
    sqlmesh_airflow = SQLMeshAirflow(engine_operator, default_catalog="spark_catalog")
  File "/opt/sqlmesh/sqlmesh/schedulers/airflow/integration.py", line 106, in __init__
    self.set_default_catalog(default_catalog)
  File "/opt/sqlmesh/sqlmesh/schedulers/airflow/integration.py", line 110, in set_default_catalog
    Variable.set(common.DEFAULT_CATALOG_VARIABLE_NAME, default_catalog)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/utils/session.py", line 71, in wrapper
    return func(*args, session=session, **kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/models/variable.py", line 177, in set
    session.flush()
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 3345, in flush
    self._flush(objects)
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 3485, in _flush
    transaction.rollback(_capture_exception=True)
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/util/langhelpers.py", line 72, in __exit__
    with_traceback=exc_tb,
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 3445, in _flush
    flush_context.execute()
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/orm/unitofwork.py", line 456, in execute
    rec.execute(self)
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/orm/unitofwork.py", line 633, in execute
    uow,
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/orm/persistence.py", line 249, in save_obj
    insert,
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/orm/persistence.py", line 1224, in _emit_insert_statements
    execution_options=execution_options,
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1614, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/sql/elements.py", line 326, in _execute_on_connection
    self, multiparams, params, execution_options
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1491, in _execute_clauseelement
    cache_hit=cache_hit,
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1846, in _execute_context
    e, statement, parameters, cursor, context
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 2027, in _handle_dbapi_exception
    sqlalchemy_exception, with_traceback=exc_info[2], from_=e
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1803, in _execute_context
    cursor, statement, parameters, context
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 719, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "variable_key_key"
DETAIL:  Key (key)=(sqlmesh_default_catalog) already exists.
```

**Pull request merged**